KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,407 @@
+            end
+        end
+    end
+    return rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, id_shape_cnt
+end
+
+
+"""
+    gradunique_amp_id, gradunique_tim_id, id_shape_cnt = get_gradunique(gradunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+
+Returns the unique shapes for the amplitude and time of the "gradunique_obj_id" vector.
+Requires an initial integer counter "id_shape_cnt" to asign IDs incrementally.
+"""
+function get_gradunique(gradunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+    # Find shapes for magnitude and time gradients
+    gradunique_amp_id, gradunique_tim_id = [], []
+    for (obj,_) ∈ gradunique_obj_id
+        shape_amp = obj.A / magsign(obj.A)
+        if all([!(length(shape_amp) == length(shape_amp_unique) && shape_amp ≈ shape_amp_unique) for (shape_amp_unique,_) ∈ gradunique_amp_id])
+            push!(gradunique_amp_id, [shape_amp, id_shape_cnt])
+            id_shape_cnt = id_shape_cnt + 1
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T])/seq.DEF["GradientRasterTime"]
+            if all([!(length(shape_tim) == length(shape_tim_unique) && shape_tim ≈ shape_tim_unique) for (shape_tim_unique,_) ∈ gradunique_tim_id])
+                push!(gradunique_tim_id, [shape_tim, id_shape_cnt])
+                id_shape_cnt = id_shape_cnt + 1
+            end
+        end
+    end
+    return gradunique_amp_id, gradunique_tim_id, id_shape_cnt
+end
+
+"""
+    compressed_data = compress(data::Vector)
+
+Returns the compressed data vector `compressed_data` for the intupt vector `data` using the
+algorithm for encoding the derivative in a run-length compressed format.
+"""
+function compress(data)
+    return compress_shape(data)[2]
+end
+
+"""
+    output = compress_if_necessary(data::Vector)
+
+Returns the vector of compressed data as long as there is a real compression. Otherwise it
+resturns the original data vector.
+"""
+function compress_if_necessary(data)
+    data_compressed = compress(data)
+    return length(data_compressed) == length(data) ? data : data_compressed
+end
+
+"""
+    write_seq(seq::Sequence, filename::String)
+
+Writes a .seq file for a given sequence `seq` y the location `filename`
+"""
+function write_seq(seq::Sequence, filename)
+
+    # Get the unique objects (RF, Grad y ADC) and its IDs
+    rfunique_obj_id  = get_typeunique_obj_id(get_typeon_obj(seq, "rf"))
+    grunique_obj_id  = get_typeunique_obj_id(get_typeon_obj(seq, "gr"))
+    adcunique_obj_id = get_typeunique_obj_id(get_typeon_obj(seq, "adc"))
+    gradunique_obj_id = [[obj, id] for (obj, id) ∈ grunique_obj_id if length(obj.A) != 1]
+    trapunique_obj_id = [[obj, id] for (obj, id) ∈ grunique_obj_id if length(obj.A) == 1]
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, id_shape_cnt = get_rfunique(rfunique_obj_id, 1, seq)
+    gradunique_amp_id, gradunique_tim_id, _ = get_gradunique(gradunique_obj_id, id_shape_cnt, seq)
```

Comment:
Arent all ids unique?

Reply:
I renamed variables in  2dd30b0 bcf5515 511dc91 72718e6